{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.7.8\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "PIL                 8.1.2\n",
      "anyio               NA\n",
      "attr                20.3.0\n",
      "babel               2.9.0\n",
      "backcall            0.2.0\n",
      "brotli              NA\n",
      "cairo               1.20.0\n",
      "certifi             2020.12.05\n",
      "cffi                1.14.5\n",
      "chardet             4.0.0\n",
      "cloudpickle         1.6.0\n",
      "colorama            0.4.4\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "cytoolz             0.11.0\n",
      "dask                2021.03.1\n",
      "dateutil            2.8.1\n",
      "decorator           4.4.2\n",
      "fsspec              0.7.4\n",
      "google              NA\n",
      "h5py                2.10.0\n",
      "idna                2.10\n",
      "igraph              0.8.3\n",
      "ipykernel           5.5.0\n",
      "ipython_genutils    0.2.0\n",
      "jedi                0.18.0\n",
      "jinja2              2.11.3\n",
      "joblib              1.0.1\n",
      "json5               NA\n",
      "jsonschema          3.2.0\n",
      "jupyter_server      1.4.1\n",
      "jupyterlab_server   2.3.0\n",
      "kiwisolver          1.3.1\n",
      "leidenalg           0.8.3\n",
      "llvmlite            0.34.0\n",
      "louvain             0.7.0\n",
      "lpy                 NA\n",
      "markupsafe          1.1.1\n",
      "matplotlib          3.6.1\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "nbclassic           NA\n",
      "nbformat            5.1.2\n",
      "numba               0.51.2\n",
      "numpy               1.19.5\n",
      "packaging           20.9\n",
      "pandas              1.1.4\n",
      "parso               0.8.1\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prometheus_client   NA\n",
      "prompt_toolkit      3.0.16\n",
      "psutil              5.8.0\n",
      "ptyprocess          0.7.0\n",
      "pvectorc            NA\n",
      "pyarrow             0.16.0\n",
      "pygments            2.8.0\n",
      "pyparsing           2.4.7\n",
      "pyrsistent          NA\n",
      "pytz                2021.1\n",
      "requests            2.25.1\n",
      "rpy2                3.4.3\n",
      "ruamel              NA\n",
      "scipy               1.6.1\n",
      "send2trash          NA\n",
      "session_info        1.0.0\n",
      "setuptools_scm      NA\n",
      "six                 1.15.0\n",
      "sklearn             0.22\n",
      "sniffio             1.2.0\n",
      "socks               1.7.1\n",
      "sphinxcontrib       NA\n",
      "storemagic          NA\n",
      "tblib               1.7.0\n",
      "texttable           1.6.3\n",
      "tlz                 0.11.0\n",
      "toolz               0.11.1\n",
      "tornado             6.1\n",
      "traitlets           5.0.5\n",
      "typing_extensions   NA\n",
      "tzlocal             NA\n",
      "urllib3             1.26.3\n",
      "wcwidth             0.2.5\n",
      "yaml                5.4.1\n",
      "zmq                 22.0.3\n",
      "-----\n",
      "IPython             7.21.0\n",
      "jupyter_client      6.1.11\n",
      "jupyter_core        4.7.1\n",
      "jupyterlab          3.0.9\n",
      "notebook            6.2.0\n",
      "-----\n",
      "Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27) [GCC 9.3.0]\n",
      "Linux-4.15.0-112-generic-x86_64-with-glibc2.10\n",
      "-----\n",
      "Session information updated at 2022-10-20 14:31\n"
     ]
    }
   ],
   "source": [
    "# Integration of biopsies and organ donors from 2 studies\n",
    "import anndata, numpy as np, scipy, pandas as pd, scanpy as sc, imp, re, rpy2\n",
    "import lpy # this is a local file\n",
    "%load_ext rpy2.ipython\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mCreate AnnData Object\u001b[0m\u001b[34m\n",
      "def createAnnData(folderlist, prefix, souporcell_folderlist = None, souporcell_genodico = None, autoinclude=[\"percent_mito\", \"log2p1_count\", \"n_genes\"], min_cell_per_gene_allowed=3, min_gene_per_cell_allowed=500, min_umi_per_cell_allowed=None, sample_obskey = \"sample_names\",doqcplots=False, mitogeneprefix=\"MT-\", exclude_genotypic_doublets= True, cellbarcode_blacklist = None, gex_only=True,genefilter_list = None, geneconcatdico=None, maxcellperob = None, do_log2_normalize=True, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mCreate AnnData Object\\033[0m\\033[34m\"); print(inspect.getsource(createAnnData));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    adatas = []\n",
      "    from scipy import sparse\n",
      "    if cellbarcode_blacklist is not None: cellbarcode_blacklist = {x:0 for x in cellbarcode_blacklist}\n",
      "    \n",
      "    \n",
      "\n",
      "    for i in range(len(folderlist)):\n",
      "        print(\"Processing \" + prefix[i])\n",
      "        adatas.append(sc.read_10x_mtx(folderlist[i],gex_only =gex_only))\n",
      "        if genefilter_list is not None:\n",
      "            gflt = np.array([x in genefilter_list for x in adatas[-1].var_names])\n",
      "            adatas[-1] = anndata.AnnData(X=adatas[-1].X[:,gflt],var=adatas[-1].var[gflt],obs=adatas[-1].obs)\n",
      "        if geneconcatdico is not None:\n",
      "            if (i == 0):\n",
      "                from scipy.sparse import coo_matrix, hstack\n",
      "                remap = {}\n",
      "                for x,y in geneconcatdico.items():\n",
      "                    for t in y: remap[t] = x\n",
      "                collectionkeys = [x for x in geneconcatdico.keys()] \n",
      "            vard = adatas[-1].var[[x not in remap for x in adatas[-1].var_names]]\n",
      "            summat = np.ndarray([adatas[-1].shape[0], len(geneconcatdico)], dtype = 'float')\n",
      "            for x in range(len(collectionkeys)):\n",
      "                vard.loc[collectionkeys[x]] = [collectionkeys[x], \"collection\"]\n",
      "                summat[:,x] = np.sum(adatas[-1].X[:,[False if c not in remap else remap[c] == collectionkeys[x] for c in adatas[-1].var_names] ], axis =1).reshape((-1))\n",
      "            adatas[-1] = anndata.AnnData(hstack([adatas[-1].X[:, [x not in remap for x in adatas[-1].var_names] ], coo_matrix(summat) ]).tocsr(), obs = adatas[-1].obs, var = vard)\n",
      "\n",
      "        if min_umi_per_cell_allowed is not None: sc.pp.filter_cells(adatas[-1], min_counts= min_umi_per_cell_allowed, inplace=True)\n",
      "        print( str(len(adatas[i].obs_names)) + \" cells found\")\n",
      "        if souporcell_folderlist is not None:\n",
      "            try:\n",
      "                res = pd.read_csv(souporcell_folderlist[i] + \"clusters.tsv\",sep='\\t')\n",
      "                #if (res.shape[0] == len(adatas[i].obs_names)): adatas[i].obs[\"souporcell\"] = res.status.to_list() # same barcode list\n",
      "                #else:\n",
      "                namemap = {res[\"barcode\"][x]  : x for x in range(res.shape[0]) }\n",
      "                adatas[i].obs[\"souporcell\"] = [ res.status[namemap[x]] if x in namemap else \"unassigned\" for x in adatas[i].obs_names]\n",
      "                if exclude_genotypic_doublets:\n",
      "                    adatas[i] = adatas[i][[x not in [\"unassigned\", \"doublets\"] for x in  adatas[i].obs[\"souporcell\"] ]]\n",
      "                adatas[i].obs[\"demultiplexed\"] = [ res.assignment[namemap[x]] if x in namemap else \"unassigned\" for x in adatas[i].obs_names]\n",
      "                #adatas[i].obs[\"souporcell\"] = res.status.to_list()\n",
      "                if (souporcell_genodico is None) or (prefix[i] not in souporcell_genodico.keys()):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [prefix[i] + \"_genotype_\" + str(x) if y == \"singlet\" else y for x,y in zip( adatas[i].obs[\"demultiplexed\"] , adatas[i].obs[\"souporcell\"]) ]\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [ souporcell_genodico[prefix[i]][int(x)] if y == \"singlet\" else y for x,y in zip( adatas[i].obs[\"demultiplexed\"] , adatas[i].obs[\"souporcell\"]) ]\n",
      "            except:\n",
      "                print(\"No valid genotyping data for \"+ str(len(adatas[i].obs_names)) + \" cells within \" + prefix[i] + \"! Setting everything to singlets\")\n",
      "                adatas[i].obs[\"souporcell\"] = \"singlet\"\n",
      "                if (souporcell_genodico is None):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = prefix[i] + \"_genotype_0\"\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = souporcell_genodico[prefix[i]][0]\n",
      "        adatas[i].obs[sample_obskey] = [prefix[i] for hehe in adatas[i].obs_names]\n",
      "        adatas[i].obs_names = [prefix[i] + \"_\" + x for x in adatas[i].obs_names]\n",
      "        if (cellbarcode_blacklist is not None): adatas[i] = adatas[i][[x not in cellbarcode_blacklist for x in adatas[i].obs_names]]\n",
      "        if \"n_count\" in autoinclude:\n",
      "            adatas[i].obs['n_count'] = np.sum(adatas[i].X, axis=1).A\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            adatas[i].obs['log2p1_count'] = pd.to_numeric(np.log1p(np.sum(adatas[i].X, axis=1).A1) / math.log(2), downcast='float')\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            mito_genes = [name for name in adatas[i].var_names if name.startswith(mitogeneprefix)]\n",
      "            adatas[i].obs['percent_mito'] = pd.to_numeric(np.sum(adatas[i][:, mito_genes].X, axis=1).A1 / np.sum(adatas[i].X, axis=1).A1, downcast='float')\n",
      "        if \"rpmask\" in autoinclude:\n",
      "            rpgenes = [name for name in adatas[i].var_names if name.startswith(\"RPM_\")]\n",
      "            adatas[i].obs['percent_rpmasked'] = pd.to_numeric(np.sum(adatas[i][:, rpgenes].X, axis=1).A1 / np.sum(adatas[i].X, axis=1).A1, downcast='float')\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            adatas[i].obs['n_genes'] = np.sum(adatas[i].X != 0, axis=1).A1\n",
      "        \n",
      "        adatas[i].obs[sample_obskey].value_counts()\n",
      "    adata = mergeAnnData(adatas,index_unique = None,maxcellperob= maxcellperob)\n",
      "\n",
      "    if min_gene_per_cell_allowed is not None: sc.pp.filter_cells(adata, min_genes=min_gene_per_cell_allowed, inplace=True)\n",
      "    if min_cell_per_gene_allowed is not None: sc.pp.filter_genes(adata, min_cells=min_cell_per_gene_allowed, inplace=True)\n",
      "    if doqcplots is True:\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['log2p1_count'], groupby=sample_obskey, rotation=90)\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['n_genes'], groupby=sample_obskey, rotation=90)\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['percent_mito'], groupby=sample_obskey, rotation=90)\n",
      "    if do_log2_normalize:\n",
      "        adata.raw = anndata.AnnData(adata.X, var = adata.var)\n",
      "        adata.X.data = np.log1p(adata.X.data) / math.log(2)\n",
      "        print(adata.X.indptr)\n",
      "    return adata;\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "Processing T1_C-0\n",
      "10830 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing T2_C-0\n",
      "12636 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing T3_C-0\n",
      "20277 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing I-O-5_8_C-0\n",
      "16743 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing I-O-5_8-N_C-0\n",
      "15920 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing I-O-6_9_C-0\n",
      "13140 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing I-O-6_9-N_C-0\n",
      "15815 cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing I-O-7_C-0\n",
      "13153 cells found\n",
      "No valid genotyping data for 13153 cells within I-O-7_C-0! Setting everything to singlets\n",
      "Processing I-O-7-N_C-0\n",
      "19404 cells found\n",
      "No valid genotyping data for 19404 cells within I-O-7-N_C-0! Setting everything to singlets\n",
      "Processing hani\n",
      "97039 cells found\n",
      "No valid genotyping data for 97039 cells within hani! Setting everything to singlets\n",
      "Processing S-HC008\n",
      "3673 cells found\n",
      "No valid genotyping data for 3673 cells within S-HC008! Setting everything to singlets\n",
      "Processing S-HC010\n",
      "5138 cells found\n",
      "No valid genotyping data for 5138 cells within S-HC010! Setting everything to singlets\n",
      "Processing S-HC013\n",
      "10477 cells found\n",
      "No valid genotyping data for 10477 cells within S-HC013! Setting everything to singlets\n",
      "Processing S-HC015\n",
      "10772 cells found\n",
      "No valid genotyping data for 10772 cells within S-HC015! Setting everything to singlets\n",
      "Processing S-HC017\n",
      "10067 cells found\n",
      "No valid genotyping data for 10067 cells within S-HC017! Setting everything to singlets\n",
      "Processing S-HC014\n",
      "8393 cells found\n",
      "No valid genotyping data for 8393 cells within S-HC014! Setting everything to singlets\n",
      "Processing S-M044-1\n",
      "5700 cells found\n",
      "No valid genotyping data for 5700 cells within S-M044-1! Setting everything to singlets\n",
      "Processing S-M046\n",
      "4235 cells found\n",
      "No valid genotyping data for 4235 cells within S-M046! Setting everything to singlets\n",
      "Processing S-M047\n",
      "4172 cells found\n",
      "No valid genotyping data for 4172 cells within S-M047! Setting everything to singlets\n",
      "Processing S-M067\n",
      "9958 cells found\n",
      "No valid genotyping data for 9958 cells within S-M067! Setting everything to singlets\n",
      "Processing S-M044-2\n",
      "5685 cells found\n",
      "No valid genotyping data for 5685 cells within S-M044-2! Setting everything to singlets\n",
      "Processing S-M050\n",
      "3821 cells found\n",
      "No valid genotyping data for 3821 cells within S-M050! Setting everything to singlets\n",
      "Processing S-M051\n",
      "5096 cells found\n",
      "No valid genotyping data for 5096 cells within S-M051! Setting everything to singlets\n",
      "Processing S-M052\n",
      "4987 cells found\n",
      "No valid genotyping data for 4987 cells within S-M052! Setting everything to singlets\n",
      "Processing S-M055\n",
      "4117 cells found\n",
      "No valid genotyping data for 4117 cells within S-M055! Setting everything to singlets\n",
      "Processing S-S054\n",
      "5763 cells found\n",
      "No valid genotyping data for 5763 cells within S-S054! Setting everything to singlets\n",
      "Processing S-S057\n",
      "4544 cells found\n",
      "No valid genotyping data for 4544 cells within S-S057! Setting everything to singlets\n",
      "Processing S-S061\n",
      "2863 cells found\n",
      "No valid genotyping data for 2863 cells within S-S061! Setting everything to singlets\n",
      "Processing S-S055\n",
      "5103 cells found\n",
      "No valid genotyping data for 5103 cells within S-S055! Setting everything to singlets\n",
      "Processing S-S059\n",
      "3772 cells found\n",
      "No valid genotyping data for 3772 cells within S-S059! Setting everything to singlets\n",
      "Processing S-S062\n",
      "2884 cells found\n",
      "No valid genotyping data for 2884 cells within S-S062! Setting everything to singlets\n",
      "Processing S-S064\n",
      "2579 cells found\n",
      "No valid genotyping data for 2579 cells within S-S064! Setting everything to singlets\n",
      "Processing S-S065\n",
      "8129 cells found\n",
      "No valid genotyping data for 8129 cells within S-S065! Setting everything to singlets\n",
      "Processing S-S067\n",
      "5592 cells found\n",
      "No valid genotyping data for 5592 cells within S-S067! Setting everything to singlets\n",
      "Processing S-S068\n",
      "5298 cells found\n",
      "No valid genotyping data for 5298 cells within S-S068! Setting everything to singlets\n",
      "[        0      1319      3104 ... 622796485 622797537 622798948]\n",
      "[        0      1319      3104 ... 621225705 621226757 621228168]\n"
     ]
    }
   ],
   "source": [
    "#list sample and load metadata\n",
    "meta = pd.read_csv(\"./sample_meta_n.tsv\",sep='\\t')\n",
    "\n",
    "# define path the location of the count matrices for every sample listed in the metadata file loaded\n",
    "folderbase = {\"Javi\" : \"/lustre/scratch117/cellgen/team292/lh20/\", \"Hani\" : \"/lustre/scratch117/cellgen/team292/lh20/javi/\", \"China\" : \"/lustre/scratch117/cellgen/team292/lh20/javi/\"}\n",
    "folderinner = {\"Javi\" : \"/outs/filtered_feature_bc_matrix/\", \"Hani\" : \"/\", \"China\" : \"/\"}\n",
    "foldersoupinner = {\"Javi\" : \"/outs/filtered_feature_bc_matrix/\", \"Hani\" : \"/\", \"China\" : \"/\"}\n",
    "\n",
    "folderlist = [folderbase[meta[\"StudyName\"][i]] + meta[\"FolderName\"][i] + folderinner[meta[\"StudyName\"][i]] for i in range(meta.shape[0]) ] \n",
    "soupxfolderlist = [folderbase[meta[\"StudyName\"][i]] + meta[\"FolderName\"][i] + foldersoupinner[meta[\"StudyName\"][i]] for i in range(meta.shape[0]) ] \n",
    "souporcell_genodico = {meta[\"FolderName\"][i] : [meta[\"genotype0\"][i], meta[\"genotype1\"][i],meta[\"genotype2\"][i], meta[\"genotype3\"][i]] for i in range(meta.shape[0])}\n",
    "\n",
    "adata = lpy.createAnnData(folderlist, meta[\"FolderName\"], souporcell_folderlist = soupxfolderlist, souporcell_genodico = souporcell_genodico, mitogeneprefix= \"GRCh38___MT-\", min_cell_per_gene_allowed=0, doqcplots=False, gex_only= False, doinspect=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata2 = anndata.read_h5ad(\"../../covid_portal_210320_with_raw.h5ad\")\n",
    "import math\n",
    "subm = adata2.layers[\"raw\"][[x for x in dadico.values() ], 24737:24929 ]\n",
    "res = {\"hani_\" + adata2.obs_names[x] : x for x in range(len(adata2.obs_names))}\n",
    "dadico = {x : res[adata.obs_names[x]] for x in range(len(adata.obs_names)) if adata.obs_names[x] in res}\n",
    "adata.X[[x for x in dadico.keys()], 33567:33759] = subm\n",
    "del adata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmpraw = anndata.AnnData(X= adata.raw.X, var=adata.raw.var)\n",
    "tmpraw.var_names = [re.sub(\"AB_GRCh38___\", \"\",re.sub(\"^\", \"AB_\",x)) for x in tmpraw.var_names]\n",
    "adata.raw = tmpraw\n",
    "adata.var_names = [re.sub(\"AB_GRCh38___\", \"\",re.sub(\"^\", \"AB_\",x)) for x in adata.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "I-O-7-N_C-0      19043\n",
       "T3_C-0           19025\n",
       "I-O-5_8_C-0      16193\n",
       "I-O-5_8-N_C-0    15339\n",
       "I-O-6_9-N_C-0    15169\n",
       "I-O-7_C-0        12892\n",
       "I-O-6_9_C-0      12620\n",
       "T2_C-0           12010\n",
       "MH8919333        11970\n",
       "S-HC015          10771\n",
       "S-HC013          10477\n",
       "T1_C-0           10177\n",
       "S-HC017          10065\n",
       "S-M067            9952\n",
       "S-HC014           8393\n",
       "S-S065            8129\n",
       "MH8919332         7850\n",
       "MH8919178         6948\n",
       "newcastle74       6850\n",
       "BGCV01_CV0904     6847\n",
       "MH8919282         6802\n",
       "MH8919283         6642\n",
       "S-S054            5763\n",
       "S-M044-1          5700\n",
       "S-M044-2          5684\n",
       "newcastle65       5604\n",
       "S-S067            5592\n",
       "S-S068            5298\n",
       "S-HC010           5138\n",
       "S-S055            5103\n",
       "S-M051            5096\n",
       "S-M052            4987\n",
       "S-S057            4544\n",
       "S-M046            4235\n",
       "MH8919179         4214\n",
       "S-M047            4172\n",
       "S-M055            4117\n",
       "S-M050            3821\n",
       "S-S059            3772\n",
       "S-HC008           3673\n",
       "BGCV13_CV0934     3408\n",
       "MH8919176         3085\n",
       "BGCV14_CV0940     2957\n",
       "BGCV09_CV0917     2923\n",
       "S-S062            2884\n",
       "S-S061            2861\n",
       "BGCV12_CV0926     2620\n",
       "BGCV02_CV0902     2598\n",
       "S-S064            2579\n",
       "MH8919177         2175\n",
       "MH8919227         1990\n",
       "BGCV08_CV0915     1938\n",
       "BGCV04_CV0911     1905\n",
       "BGCV10_CV0939     1718\n",
       "BGCV05_CV0929     1465\n",
       "BGCV01_CV0902     1281\n",
       "MH8919226         1124\n",
       "BGCV15_CV0944      634\n",
       "Name: sample_names, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hameta = pd.read_csv(\"hani_meta.csv\")\n",
    "adata.obs[\"tmp\"] = [re.sub(\"^hani_\", \"\", x) for x in adata.obs.index]\n",
    "adata = lpy.addMetadata(adata, hameta, \"tmp\", \"covid_index\", column_list= [\"sample_id\", \"full_clustering\",\"initial_clustering\", \"patient_id\"])\n",
    "del adata.obs[\"tmp\"]\n",
    "\n",
    "adata.obs[\"sample_names\"] = [y if x == \"hani\" else x for  x,y in zip(adata.obs[\"sample_names\"],adata.obs[\"sample_id\"])]\n",
    "adata.obs[\"demultiplexed\"] = [y if x == \"MANY\" else x for  x,y in zip(adata.obs[\"demultiplexed\"],adata.obs[\"sample_id\"])]\n",
    "\n",
    "del adata.obs[\"sample_id\"]\n",
    "adata.obs[\"sample_names\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "China      142806\n",
       "Javi       132468\n",
       "Haniffa     95548\n",
       "Name: StudyName, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = lpy.addMetadata(adata, meta, \"sample_names\", \"FolderName\", column_list= meta.columns[ [\"genotype\" not in x for x in meta.columns ] ])\n",
    "adata.obs[\"StudyName\"] = [\"Haniffa\" if x ==\"\" else x for  x in adata.obs[\"StudyName\"]]\n",
    "adata.obs[\"StudyName\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CVID1            60531\n",
      "CVID2            27393\n",
      "CVID4            14473\n",
      "doublet          12518\n",
      "MH8919333        11970\n",
      "CVID3            11565\n",
      "S-M044           11384\n",
      "S-HC015          10771\n",
      "S-HC013          10477\n",
      "S-HC017          10065\n",
      "S-M067            9952\n",
      "S-HC014           8393\n",
      "S-S065            8129\n",
      "MH8919332         7850\n",
      "MH8919178         6948\n",
      "newcastle74       6850\n",
      "BGCV01_CV0904     6847\n",
      "MH8919282         6802\n",
      "MH8919283         6642\n",
      "CVID5             5988\n",
      "S-S054            5763\n",
      "newcastle65       5604\n",
      "S-S067            5592\n",
      "S-S068            5298\n",
      "S-HC010           5138\n",
      "S-S055            5103\n",
      "S-M051            5096\n",
      "S-M052            4987\n",
      "S-S057            4544\n",
      "S-M046            4235\n",
      "MH8919179         4214\n",
      "S-M047            4172\n",
      "S-M055            4117\n",
      "S-M050            3821\n",
      "S-S059            3772\n",
      "S-HC008           3673\n",
      "BGCV13_CV0934     3408\n",
      "MH8919176         3085\n",
      "BGCV14_CV0940     2957\n",
      "BGCV09_CV0917     2923\n",
      "S-S062            2884\n",
      "S-S061            2861\n",
      "BGCV12_CV0926     2620\n",
      "BGCV02_CV0902     2598\n",
      "S-S064            2579\n",
      "MH8919177         2175\n",
      "MH8919227         1990\n",
      "BGCV08_CV0915     1938\n",
      "BGCV04_CV0911     1905\n",
      "BGCV10_CV0939     1718\n",
      "BGCV05_CV0929     1465\n",
      "BGCV01_CV0902     1281\n",
      "MH8919226         1124\n",
      "BGCV15_CV0944      634\n",
      "Name: demultiplexed, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "genometa = pd.read_csv(\"genotype.tsv\",sep='\\t')\n",
    "pd.set_option('max_rows', 700)\n",
    "print(adata.obs[\"demultiplexed\"].value_counts())\n",
    "adata = lpy.addMetadata(adata, genometa, \"demultiplexed\", \"Patients\", column_list= [\"Age\", \"Sex\",\"SARS-CoV-2\", \"Comorbidities\", \"Medication\"], doinspect =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mDetect Doublets and defining cells to filter\u001b[0m\u001b[34m\n",
      "def scrub(adata, batch_obsattrib, bonf_threshold = 0.01, add_qc_metrics=False,mito_prefix= \"MT-\", obskey_cellfilter = \"filtered_cells\", add_cell_filter={\"max_percent_mito\": 0.15, \"scrublet_local_pred\": False}, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mDetect Doublets and defining cells to filter\\033[0m\\033[34m\"); print(inspect.getsource(scrub));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    \n",
      "    import scrublet as scr\n",
      "    import scanpy as sc\n",
      "    print(\"spliting data using attribute \" + batch_obsattrib)\n",
      "    adatas = splitAnnData(adata, batch_obsattrib)\n",
      "\n",
      "    if (add_qc_metrics):\n",
      "        mito_genes = [name for name in adata.var_names if name.startswith(mito_prefix)]\n",
      "        adata.obs['log2p1_RNA_count'] = np.log1p(adata.X.sum(axis=1).A1) / math.log(2)\n",
      "        adata.obs['percent_mito'] = np.sum(adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
      "\n",
      "    dalist = list(adata.obs_names)\n",
      "    tmap = {}\n",
      "    for i in range(len(adata.obs_names)):\n",
      "        tmap.update( {adata.obs_names[i] : i})\n",
      "    adata.obs[\"scrublet_pred\"] = False\n",
      "    adata.obs[\"scrublet_local_pred\"] = False\n",
      "    adata.obs[\"scrublet_score\"] = 0.0\n",
      "    adata.obs[\"scrublet_cluster_score\"] = 0.0\n",
      "\n",
      "    # Luz's double clustering approach\n",
      "    for i in range(len(adatas)):\n",
      "\n",
      "        print(\"processing \" + str(i) + \"/\" + str(len(adatas)))\n",
      "        curscr = scr.Scrublet(adatas[i].X)\n",
      "        doublet_scores, predicted_doublets = curscr.scrub_doublets()\n",
      "        adatas[i].obs['scrublet_score'] = doublet_scores\n",
      "        adatas[i].obs['scrublet_pred'] = predicted_doublets\n",
      "        #overcluster prep. run turbo basic scanpy pipeline\n",
      "        sc.pp.normalize_per_cell(adatas[i], counts_per_cell_after=1e4)\n",
      "        sc.pp.log1p(adatas[i])\n",
      "        sc.pp.highly_variable_genes(adatas[i], min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
      "        adatas[i] = adatas[i][:, adatas[i].var['highly_variable']]\n",
      "        sc.pp.scale(adatas[i], max_value=10)\n",
      "        sc.tl.pca(adatas[i], svd_solver='arpack')\n",
      "        sc.pp.neighbors(adatas[i])\n",
      "        #overclustering proper - do basic clustering first, then cluster each cluster\n",
      "        sc.tl.leiden(adatas[i])\n",
      "        adatas[i].obs['leiden'] = [str(i) for i in adatas[i].obs['leiden']]\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adata_sub = adatas[i][adatas[i].obs['leiden']==clus].copy()\n",
      "            sc.tl.leiden(adata_sub)\n",
      "            adata_sub.obs['leiden'] = [clus+','+i for i in adata_sub.obs['leiden']]\n",
      "            adatas[i].obs.loc[adata_sub.obs_names,'leiden'] = adata_sub.obs['leiden']\n",
      "\n",
      "        #compute the cluster scores - the median of Scrublet scores per overclustered cluster\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_cluster_score'] = \\\n",
      "                np.median(adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_score'])\n",
      "        #now compute doublet p-values. figure out the median and mad (from above-median values) for the distribution\n",
      "        med = np.median(adatas[i].obs['scrublet_cluster_score'])\n",
      "        mask = adatas[i].obs['scrublet_cluster_score']>med\n",
      "        mad = np.median(adatas[i].obs['scrublet_cluster_score'][mask]-med)\n",
      "        #let's do a one-sided test. the Bertie write-up does not address this but it makes sense\n",
      "        zscores = (adatas[i].obs['scrublet_cluster_score'].values - med) / (1.4826 * mad)\n",
      "        #adatas[i].obs['zscore'] = zscores\n",
      "        pvals = 1-scipy.stats.norm.cdf(zscores)\n",
      "        #adatas[i].obs['bh_pval'] = bh(pvals)\n",
      "        pvals = bonf(pvals)\n",
      "        adatas[i].obs['scrublet_local_pred'] = pvals < bonf_threshold\n",
      "        map = [tmap[s] for s in adatas[i].obs_names]\n",
      "        print(\"annoying values\")\n",
      "        for j in range(len(adatas[i].obs_names)):\n",
      "            adata.obs[\"scrublet_pred\" ][map[j]] = adatas[i].obs[\"scrublet_pred\"][j]\n",
      "            adata.obs[\"scrublet_local_pred\"][map[j]] = adatas[i].obs[\"scrublet_local_pred\"][j]\n",
      "            adata.obs[\"scrublet_cluster_score\"][map[j]] =  adatas[i].obs[\"scrublet_cluster_score\"][j]\n",
      "            adata.obs[\"scrublet_score\"][map[j]] =  pvals[j]\n",
      "\n",
      "    adata.obs[\"scrublet_cluster_score\"] = pd.to_numeric(adata.obs[\"scrublet_cluster_score\"], downcast='float')\n",
      "    adata.obs[\"scrublet_score\"] = pd.to_numeric(adata.obs[\"scrublet_score\"], downcast='float')\n",
      "\n",
      "    if add_cell_filter is not None:\n",
      "        curflt = np.zeros(adata.obs.shape[0], dtype=bool) \n",
      "        for k,v in add_cell_filter.items():\n",
      "            if k == \"max_percent_mito\":\n",
      "                print(str(sum(adata.obs['percent_mito'] > v)) + \" cells filtered by mito threshold\")\n",
      "                curflt |= adata.obs['percent_mito'] > v\n",
      "            else:\n",
      "                print(str(sum(adata.obs[k] != v)) + \" cells filtered by \" + k + \" == \" + str(v) + \" criterion\")\n",
      "                curflt |= (adata.obs[k] != v)\n",
      "        adata.obs[obskey_cellfilter] = curflt\n",
      "    return adata\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "spliting data using attribute sample_names\n",
      "processing 0/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 1.0%\n",
      "Estimated detectable doublet fraction = 32.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.1%\n",
      "Elapsed time: 12.8 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: NumExpr detected 26 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annoying values\n",
      "processing 1/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.39\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 65.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 13.1 seconds\n",
      "annoying values\n",
      "processing 2/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 32.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.9%\n",
      "Elapsed time: 35.1 seconds\n",
      "annoying values\n",
      "processing 3/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 25.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 2.8 seconds\n",
      "annoying values\n",
      "processing 4/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 46.0%\n",
      "Estimated detectable doublet fraction = 92.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 50.0%\n",
      "Elapsed time: 6.6 seconds\n",
      "annoying values\n",
      "processing 5/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.36\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 2.5 seconds\n",
      "annoying values\n",
      "processing 6/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 65.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 10.0 seconds\n",
      "annoying values\n",
      "processing 7/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.61\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 52.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 4.9 seconds\n",
      "annoying values\n",
      "processing 8/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 68.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.4%\n",
      "Elapsed time: 21.7 seconds\n",
      "annoying values\n",
      "processing 9/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 35.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.3%\n",
      "Elapsed time: 3.4 seconds\n",
      "annoying values\n",
      "processing 10/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 62.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 9.9 seconds\n",
      "annoying values\n",
      "processing 11/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 2.9 seconds\n",
      "annoying values\n",
      "processing 12/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 32.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.3%\n",
      "Elapsed time: 20.9 seconds\n",
      "annoying values\n",
      "processing 13/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 38.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 11.9 seconds\n",
      "annoying values\n",
      "processing 14/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 55.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.2%\n",
      "Elapsed time: 4.0 seconds\n",
      "annoying values\n",
      "processing 15/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 57.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 1.9 seconds\n",
      "annoying values\n",
      "processing 16/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.03\n",
      "Detected doublet rate = 85.9%\n",
      "Estimated detectable doublet fraction = 98.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 87.4%\n",
      "Elapsed time: 1.6 seconds\n",
      "annoying values\n",
      "processing 17/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 63.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.0%\n",
      "Elapsed time: 4.7 seconds\n",
      "annoying values\n",
      "processing 18/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 55.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 3.7 seconds\n",
      "annoying values\n",
      "processing 19/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 1.7%\n",
      "Estimated detectable doublet fraction = 35.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 24.1 seconds\n",
      "annoying values\n",
      "processing 20/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 57.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.2%\n",
      "Elapsed time: 6.8 seconds\n",
      "annoying values\n",
      "processing 21/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 51.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.5%\n",
      "Elapsed time: 1.3 seconds\n",
      "annoying values\n",
      "processing 22/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 56.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.9%\n",
      "Elapsed time: 13.6 seconds\n",
      "annoying values\n",
      "processing 23/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 43.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 2.7 seconds\n",
      "annoying values\n",
      "processing 24/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 1.7%\n",
      "Estimated detectable doublet fraction = 51.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.3%\n",
      "Elapsed time: 1.8 seconds\n",
      "annoying values\n",
      "processing 25/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 30.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.4%\n",
      "Elapsed time: 3.5 seconds\n",
      "annoying values\n",
      "processing 26/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 42.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 2.1 seconds\n",
      "annoying values\n",
      "processing 27/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 21.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.5%\n",
      "Elapsed time: 24.1 seconds\n",
      "annoying values\n",
      "processing 28/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 56.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.2%\n",
      "Elapsed time: 13.6 seconds\n",
      "annoying values\n",
      "processing 29/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 63.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 5.6 seconds\n",
      "annoying values\n",
      "processing 30/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 59.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.6%\n",
      "Elapsed time: 14.1 seconds\n",
      "annoying values\n",
      "processing 31/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.66\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 25.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.2%\n",
      "Elapsed time: 9.3 seconds\n",
      "annoying values\n",
      "processing 32/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 3.4%\n",
      "Estimated detectable doublet fraction = 50.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.7%\n",
      "Elapsed time: 34.3 seconds\n",
      "annoying values\n",
      "processing 33/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 40.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 7.5 seconds\n",
      "annoying values\n",
      "processing 34/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 52.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 5.4 seconds\n",
      "annoying values\n",
      "processing 35/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 62.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 2.4 seconds\n",
      "annoying values\n",
      "processing 36/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 62.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 7.2 seconds\n",
      "annoying values\n",
      "processing 37/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.45\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 49.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 1.4 seconds\n",
      "annoying values\n",
      "processing 38/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 41.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 14.0 seconds\n",
      "annoying values\n",
      "processing 39/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 51.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 5.0 seconds\n",
      "annoying values\n",
      "processing 40/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 53.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 11.1 seconds\n",
      "annoying values\n",
      "processing 41/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 48.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 7.1 seconds\n",
      "annoying values\n",
      "processing 42/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 56.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.2%\n",
      "Elapsed time: 7.2 seconds\n",
      "annoying values\n",
      "processing 43/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.48\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 60.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.4%\n",
      "Elapsed time: 5.3 seconds\n",
      "annoying values\n",
      "processing 44/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 59.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.5%\n",
      "Elapsed time: 9.0 seconds\n",
      "annoying values\n",
      "processing 45/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 2.4%\n",
      "Estimated detectable doublet fraction = 43.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.5%\n",
      "Elapsed time: 25.6 seconds\n",
      "annoying values\n",
      "processing 46/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 52.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.2%\n",
      "Elapsed time: 10.8 seconds\n",
      "annoying values\n",
      "processing 47/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 35.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 1.8 seconds\n",
      "annoying values\n",
      "processing 48/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 57.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.2%\n",
      "Elapsed time: 3.7 seconds\n",
      "annoying values\n",
      "processing 49/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 57.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 6.8 seconds\n",
      "annoying values\n",
      "processing 50/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 20.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 2.7 seconds\n",
      "annoying values\n",
      "processing 51/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 33.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 3.9 seconds\n",
      "annoying values\n",
      "processing 52/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 60.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 6.0 seconds\n",
      "annoying values\n",
      "processing 53/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 1.9%\n",
      "Estimated detectable doublet fraction = 46.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.0%\n",
      "Elapsed time: 16.9 seconds\n",
      "annoying values\n",
      "processing 54/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.45\n",
      "Detected doublet rate = 0.6%\n",
      "Estimated detectable doublet fraction = 35.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.8%\n",
      "Elapsed time: 0.6 seconds\n",
      "annoying values\n",
      "processing 55/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 43.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.1%\n",
      "Elapsed time: 1.9 seconds\n",
      "annoying values\n",
      "processing 56/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 45.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.9%\n",
      "Elapsed time: 1.1 seconds\n",
      "annoying values\n",
      "processing 57/58\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 1.1%\n",
      "Estimated detectable doublet fraction = 27.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.8%\n",
      "Elapsed time: 19.3 seconds\n",
      "annoying values\n",
      "12518 cells filtered by souporcell == singlet criterion\n",
      "617 cells filtered by mito threshold\n",
      "3334 cells filtered by scrublet_local_pred == False criterion\n"
     ]
    }
   ],
   "source": [
    "obskey_filteredcells = \"filtered_cells\"\n",
    "adata = lpy.scrub(adata, \"sample_names\",obskey_cellfilter= obskey_filteredcells,add_cell_filter={\"souporcell\": \"singlet\" ,\"max_percent_mito\": 0.20, \"scrublet_local_pred\": False}, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mAdd Cell Cycle annotation to anndata object\u001b[0m\u001b[34m\n",
      "def addCycleCycleAnnotation(adata, s_genes = None, g2m_genes = None, geneprefix = \"\", use_raw_data= True, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mAdd Cell Cycle annotation to anndata object\\033[0m\\033[34m\"); print(inspect.getsource(addCycleCycleAnnotation));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "\n",
      "    # uses Seurat Cell Cycles default genes by default\n",
      "    if s_genes is None: # \"MLF1IP\"\n",
      "        s_genes = [\"MCM5\",\"PCNA\",\"TYMS\",\"FEN1\",\"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\",\"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\"ATAD2\",\"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",\"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\"CHAF1B\",\"BRIP1\",\"E2F8\"]\n",
      "    if g2m_genes is None: #use default list\n",
      "        g2m_genes =[\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\"CENPF\",\"TACC3\",\"FAM64A\",\"SMC4\",\"CCNB2\",\"CKAP2L\",\"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"HN1\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\"]\n",
      "    s_genes =  [geneprefix  + x for x in s_genes]\n",
      "    g2m_genes =  [geneprefix  + x for x in g2m_genes]\n",
      "    sc.tl.score_genes_cell_cycle(adata, s_genes=s_genes, g2m_genes=g2m_genes, use_raw = use_raw_data)\n",
      "    adata.uns[\"phase_color\"] = ['#6E40AA', '#FF8C38', '#28EA8D']\n",
      "    return(adata)\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "WARNING: genes are not in var_names and ignored: ['FAM64A', 'HN1']\n"
     ]
    }
   ],
   "source": [
    "adata = lpy.addCycleCycleAnnotation(adata, doinspect =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID2\")&(adata.obs[\"sample_names\"] == \"T3_C-0\")] = \"convalescence\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID3\")&(adata.obs[\"sample_names\"] == \"T1_C-0\")] = \"baseline\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID3\")&(adata.obs[\"sample_names\"] == \"T2_C-0\")] = \"convalescence\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID3\")&(adata.obs[\"sample_names\"] == \"T3_C-0\")] = \"progression\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID4\")&(adata.obs[\"sample_names\"] == \"T1_C-0\")] = \"progression\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID4\")&(adata.obs[\"sample_names\"] == \"T2_C-0\")] = \"baseline\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID4\")&(adata.obs[\"sample_names\"] == \"T3_C-0\")] = \"convalescence\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID5\")&(adata.obs[\"sample_names\"] == \"T1_C-0\")] = \"convalescence\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID5\")&(adata.obs[\"sample_names\"] == \"T2_C-0\")] = \"progression\"\n",
    "adata.obs[\"Stage\"][ (adata.obs[\"demultiplexed\"] == \"CVID5\")&(adata.obs[\"sample_names\"] == \"T3_C-0\")] = \"baseline\"\n",
    "adata.obs[\"Sex\"] = pd.Series([\"NA\" if x is \"\" else x for x in adata.obs[\"Sex\"] ])\n",
    "adata.write_h5ad(\"int.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
